### PR TITLE
Enhances TOC entries with new features

### LINKED ISSUE #1110: Enhance summary TOC entries with links to raw results
@Anisha1234 once we have the appropriate data indexed, we'd like to enhance the summary with the following features:

- Searching and filtering the TOC tab with keywords entered by the user.
- Link the closest sample entries in the iteration table to sample graphs on the results server
- Link items in the TOC tab to their file on the results server 
- Any collapsed item should also collapse all of its children nodes in the tree view

### PR DESCRIPTION
**Pull request by Anisha1234**
Will fix #1110 

- [x] Searching the TOC tab with keywords entered by the user.

- [x] Filtering the TOC data with file extension from dropdown.

- [x] Link the closest sample entries in the iteration table to sample graphs on the results server

- [x] Link items in the TOC tab to their file on the results server

- [x] Any collapsed item should also collapse all of its children nodes in the tree view

**Preview**
1.Filtering of TOC files
main5

2.Searching name of a file
main6

3.Sample graphs integration
main4

### CODE REVIEW

**Comment by gurbirkalsi** [comment]
@Anisha1234  https://ant.design/components/table/#components-table-demo-expand-children
This example might be helpful for enabling filtering/searching that is offered by the Table component.

**Comment by portante** [comment]
@Anisha1234, can you squash these commits down to one?

**Comment by Anisha1234** [comment]
> @Anisha1234, can you squash these commits down to one?

@portante Sure, Have squashed the commits down to one.

**Comment by portante** [comment]
@Anisha1234, marked this as WIP (updating the title to match the commit title), as this is waiting for review from @gurbirkalsi first, and then we can consider merging it.

Can you also rebase this commit?

We also moved it to the 0.60 release as we are planning on pushing 0.59 out the door shortly.

**Comment by gurbirkalsi** [comment]
@Anisha1234 PR #1211 introduces a component called `TimeseriesGraph` which allows us to interact with jschart as a React component and behave according to the React lifecycle. I would suggest creating a wrapper component that queries the appropriate data and then passes that down to `TimeseriesGraph`. Anytime new data is passed to the component, reloading will be triggered by the `componentDidUpdate` listener.

File: web-server/v0.4/src/components/TableTree/index.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+      if (childRecord.name.toLowerCase().includes(value)) {
+        return true;
+      }
+      if (childRecord.children) {
+        return this.filterNames(childRecord.children, value);
+      }
+      return false;
+    });
+
+  render() {
```
Comment (gurbirkalsi) [inline web-server/v0.4/src/components/TableTree/index.js]: `TableTree` only visualizes 10 'hits' from Elasticsearch as the query size defaults to 10. 

@portante We're seeing up to 88k hits from one result parent and querying them all would be bad for performance. Is it possible to query directory information progressively as users click through different levels of the file tree?

File: web-server/v0.4/src/pages/Summary/index.js [hunk 2]
```diff
@@ -112,7 +111,48 @@ class Summary extends React.Component {
...
+    });
+  };
+
+  onload = tocTree => {
+    this.setState({
+      tocTree,
+    });
+  };
+
+  onSearchFile = value => {
```
Comment (gurbirkalsi) [inline web-server/v0.4/src/pages/Summary/index.js]: Searching for files doesn't seem to visualize the found file in the `TableTree` component. We should consider moving searching functionality to the `TableTree` component for better organization of helpers + state variables.

File: web-server/v0.4/src/components/TableTree/index.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    super(props);
+    this.fileData = [];
+    this.state = {
+      expandedKeys: [],
+    };
+  }
+
+  componentDidMount() {
+    const { dataSource, onload } = this.props;
+    onload(dataSource);
```
Comment (gurbirkalsi) [inline web-server/v0.4/src/components/TableTree/index.js]: We should consider referencing table data for `TableTree` as a component prop as opposed to managing the state in the parent component. Moving the search functionality and state to this component should help clean up this pattern.

File: web-server/v0.4/src/components/TableTree/index.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    const { expandedKeys } = this.state;
+    const configLink = `${config.config + config.controller_name}/${config.run_name}/`;
+
+    const columns = [
+      {
+        title: 'Names',
+        dataIndex: 'name',
+        key: 'name',
+        width: '60%',
+        filters: extensions,
```
Comment (gurbirkalsi) [inline web-server/v0.4/src/components/TableTree/index.js]: `extensions` visualizes the names of files as opposed to the file extension. Files processed without an extension name can be assumed as `.txt` files.

File: web-server/v0.4/src/components/TableTree/index.js (hunk 1 above)
Comment (gurbirkalsi) [reply web-server/v0.4/src/components/TableTree/index.js]: @Anisha1234 Can you see if the scan & scroll API offered by elasticsearch solves this issue for query size? https://www.elastic.co/guide/en/elasticsearch/guide/1.x/scan-scroll.html

**Comment by portante** [comment]
@Anisha1234, @gurbirkalsi, this is still marked "Work-in-Progress".  Is there any way we can get this commit rebased, refreshed, and then ready for submission?

**Comment by Anisha1234** [comment]
@gurbirkalsi @portante I have rebased the PR. The features here are already implemented perfectly in this PR but the only problem is, after changes in elastic search data indexing, the `queryResult` function in summary page is breaking as elasticsearch is returning`html` content instead of real data which need to be fixed. 
error:
```
index.js:1 uncaught at _callee3 at _callee3 
 at _callee6 
 at takeEvery(dashboard/fetchResult, _callee) 
 at _callee 
 at _callee 
 Error: TypeError: Cannot read property 'hits' of undefined
```

**Comment by Anisha1234** [comment]
@gurbirkalsi I have updated the PR and the features are completely functional. I would like to know how are we planning to proceed from here. Are we planning to restructure the ElasticSearch queries? And how to track the performance of this specific feature?

**Comment by portante** [comment]
This has been migrated to PR https://github.com/distributed-system-analysis/pbench-dashboard/pull/8.

**Pull request closed without merging**